core/src/main/java/org/elasticsearch/search/aggregations/pipeline/movavg/SimulatedAnealingMinimizer.java

```diff
@@ -0,0 +1,129 @@
+ * solution is worse, there is a temperature-dependent probability it will be randomly selected
+ * anyway.  This allows the algo to sample the problem space widely.  As iterations progress,
+ * the temperature decreases and the algorithm rejects poor solutions more regularly,
+ * theoretically honing in on a global minimum.
+ */
+public class SimulatedAnealingMinimizer {
+
+    /**
+     * Runs the simulated annealing algorithm and produces a new set of model coefficients.
+     * The method is quasi-generic, accepting any number of coefficients and trusting that
+     * the underlying model knows what to do with them
+     *
+     * @param model         The MovAvgModel to be optimized for
+     * @param train         A training set provided to the model, which predictions will be
+     *                      generated from
+     * @param test          A test set of data to compare the predictions against and derive
+     *                      a cost per set of coefficeints
+     * @return              The new set of model coefficients
+     */
+    public static MovAvgModel minimize(MovAvgModel model, EvictingQueue<Double> train, double[] test) {
+
+        double temp = 1;
+        double minTemp = 0.0001;
+        int iterations = 100;
+        double alpha = 0.9;
+
+        MovAvgModel bestModel = model;
+        MovAvgModel oldModel = model;
+        MovAvgModel newModel;
+
+        double oldCost = cost(model, train, test);
+        double bestCost = oldCost;
+        double newCost;
+        double ap;
+
+        while (temp > minTemp) {
+            for (int i = 0; i < iterations; i++) {
+                newModel = oldModel.neighboringModel();
+                newCost = cost(newModel, train, test);
+
+                ap = acceptanceProbability(oldCost, newCost, temp);
+                if (ap > Math.random()) {
+                    oldModel = newModel.clone();
+                    oldCost = newCost;
+
+                    if (newCost < bestCost) {
+                        bestCost = newCost;
+                        bestModel = newModel.clone();
+                    }
+                }
+            }
+
+            temp *= alpha;
+        }
+
+        return bestModel;
+    }
+
+    /**
+     * If the new cost is better than old, return 1.0.  Otherwise, return a double that increases
+     * as the two costs are closer to each other.
+     *
+     * @param oldCost   Old model cost
+     * @param newCost   New model cost
+     * @param temp      Current annealing temperature
+     * @return          The probability of accepting the new cost over the old
+     */
+    private static double acceptanceProbability(double oldCost, double newCost, double temp) {
+        return newCost < oldCost ? 1.0 : Math.exp(-(newCost - oldCost) / temp);
+    }
+
+    /**
+     * Calculates the "cost" of a coefficient.  E.g. how well do the new coefficients predict the
+     * training set of data when used with the model?
```

Comment:
should this not be "i.e. how well do the new coefficients predict the test set of data when used with the model and the training set of data?" ?